clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -908,311 +628,781 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+    createExprForStructOrScalar(FD);
+    createSpecialMethodCall(StreamDecl, MemberExprBases.back(), InitMethodName,
+                            FD);
+    createSpecialMethodCall(StreamDecl, MemberExprBases.back(),
+                            FinalizeMethodName, FD);
+  }
+
+  void handleSyclStreamType(const CXXBaseSpecifier &BS, QualType Ty) final {
+    // TODO: Creates init/finalize sequence and inits special sycl obj
+  }
+
+  void handlePointerType(FieldDecl *FD, QualType ArgTy) final {
+    createExprForStructOrScalar(FD);
+  }
+
+  void handleStructType(FieldDecl *FD, QualType Ty) final {
+    createExprForStructOrScalar(FD);
+  }
+
+  void handleScalarType(FieldDecl *FD, QualType Ty) final {
+    createExprForStructOrScalar(FD);
+  }
+
+  void enterStruct(const CXXRecordDecl *, FieldDecl *FD) final {
+    MemberExprBases.push_back(BuildMemberExpr(MemberExprBases.back(), FD));
+  }
+
+  void leaveStruct(const CXXRecordDecl *, const FieldDecl *FD) final {
+    MemberExprBases.pop_back();
+  }
+
+  void enterStruct(const CXXRecordDecl *RD, const CXXBaseSpecifier &BS) final {
+    // TODO : Implement
+  }
+
+  void leaveStruct(const CXXRecordDecl *RD, const CXXBaseSpecifier &BS) final {
+    // TODO : Implement
+  }
+};
+
+class SyclKernelIntHeaderCreator
+    : public SyclKernelFieldHandler<SyclKernelIntHeaderCreator> {
+  SYCLIntegrationHeader &Header;
+  const CXXRecordDecl *KernelLambda;
+  int64_t CurOffset = 0;
 
-/// Adds necessary data describing given kernel to the integration header.
-/// \param H           the integration header object
-/// \param Name        kernel name
-/// \param NameType    type representing kernel name (first template argument
-/// of single_task, parallel_for, etc)
-/// \param KernelObjTy kernel object type
-static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
-                              QualType NameType, CXXRecordDecl *KernelObjTy) {
-
-  ASTContext &Ctx = KernelObjTy->getASTContext();
-  const ASTRecordLayout &Layout = Ctx.getASTRecordLayout(KernelObjTy);
-  const std::string StableName = PredefinedExpr::ComputeName(
-      Ctx, PredefinedExpr::UniqueStableNameExpr, NameType);
-  H.startKernel(Name, NameType, StableName, KernelObjTy->getLocation());
-
-  auto populateHeaderForAccessor = [&](const QualType &ArgTy, uint64_t Offset) {
-    // The parameter is a SYCL accessor object.
-    // The Info field of the parameter descriptor for accessor contains
-    // two template parameters packed into an integer field:
-    //   - target (e.g. global_buffer, constant_buffer, local);
-    //   - dimension of the accessor.
-    const auto *AccTy = ArgTy->getAsCXXRecordDecl();
-    assert(AccTy && "accessor must be of a record type");
-    const auto *AccTmplTy = cast<ClassTemplateSpecializationDecl>(AccTy);
+  uint64_t getOffset(const CXXRecordDecl *RD) const {
+    // TODO: Figure this out! Offset of a base class.
```

Comment:
This probably should be calculated, I don't know why we didn't have any tests fail here.

Reply:
I don't think there are tests for this? This is just the offset for the base class right?

Reply:
Yes, but it ends up being a case where we capture struct B:

    struct A {
      accessor s;
    };
    struct B : A {};

We won't properly calculate the offset of 'A'.  It might work right now in this short example (since, of course, the offset is zero!).

Reply:
BTW: I'll fix this today and get a patch here somehow.  Just gotta get through all my emails first :/

Reply:
Ah, right... it is only when an accessor (maybe a stream in the future?) is a base-class.  I think I can still fix it easy enough, but a little more subtle than I thought.  It also explains why no tests found it.